File: example/Phenology_Model/Python/ShootNumber.py
'''
Created on 18 sept. 2018

@author: midingoy
'''
from math import *
def CalculateShootNumber( CanopyShootNumber,  LeafNumber,  SowingDensity, TargetFertileShoot, tilleringProfile, leafTillerNumberArray,  TillerNumber):
    
    OldCanopyShootNumber = CanopyShootNumber
        
    AverageShootNumberPerPlant = CalcShootNumber(LeafNumber,SowingDensity,TargetFertileShoot)[0]
    CanopyShootNumber = CalcShootNumber(LeafNumber,SowingDensity,TargetFertileShoot)[1]   
    
    OldtilleringProfile=list(tilleringProfile)
    leafTillerNumberArray1 = list(leafTillerNumberArray)    
        
    if (CanopyShootNumber != OldCanopyShootNumber):
        diff = CanopyShootNumber - OldCanopyShootNumber
        OldtilleringProfile.append(diff)
           
    tilleringProfile = list(OldtilleringProfile)
    
    TillerNumber = len(tilleringProfile)
        
    for i in range(len(leafTillerNumberArray1),int(ceil(LeafNumber))):
        leafTillerNumberArray1.append(TillerNumber);
    
    leafTillerNumberArray = list(leafTillerNumberArray1)
      

    return CanopyShootNumber, AverageShootNumberPerPlant, tilleringProfile, TillerNumber, leafTillerNumberArray

def Init(SowingDensity):
    CanopyShootNumber = SowingDensity
    AverageShootNumberPerPlant =1
    tilleringProfile = []
    tilleringProfile.append(SowingDensity);
    TillerNumber = 1;
    return CanopyShootNumber, AverageShootNumberPerPlant,tilleringProfile, TillerNumber
        
        
def CalcShootNumber(LeafNumber,SowingDensity,TargetFertileShoot):
    EmergedLeaves = int(max(1, ceil(LeafNumber - 1)));
    Shoots = Fibonacci(EmergedLeaves);
    CanopyShootNumber = min(Shoots * SowingDensity, TargetFertileShoot);
    AverageShootNumberPerPlant = CanopyShootNumber / SowingDensity;
    return AverageShootNumberPerPlant, CanopyShootNumber 

def Fibonacci(N):
    a = 0;
    b = 1;
    for i in range(0,N):
        temp = a;
        a = b;
        b = temp + b; 
    return a;

File: old_example/Phenology_Model/C#/ShootNumber.cs

using System;
using System.Collections.Generic;


public static class ShootNumber
{
	
	public static void CalculateShootNumber(int CanopyShootNumber, double LeafNumber, double SowingDensity,
	                                 double TargetFertileShoot, List<double>tilleringProfile, List<double>leafTillerNumberArray, int TillerNumber)
	{
		
		int OldCanopyShootNumber = CanopyShootNumber;
		CalcShootNumber(LeafNumber,SowingDensity,TargetFertileShoot);
		
		
		var OldtilleringProfile = new List<double>();
		for (int i = 0; i < tilleringProfile.Count;i++ ) OldtilleringProfile.Add(tilleringProfile[i]);
		
		var leafTillerNumberArray1 = new List<double>();
		for (int i = 0; i < leafTillerNumberArray.Count;i++ ) leafTillerNumberArray1.Add(leafTillerNumberArray[i]);
		

		if (CanopyShootNumber != OldCanopyShootNumber)
		{
			OldtilleringProfile.Add(CanopyShootNumber - OldCanopyShootNumber);
		}
		
		tilleringProfile = new List<double>();
		for (int i = 0; i < OldtilleringProfile.Count;i++ ) tilleringProfile.Add(OldtilleringProfile[i]);

		TillerNumber = tilleringProfile.Count;
		

		for (int i = leafTillerNumberArray1.Count; i < Math.Ceiling(LeafNumber); i++)
		{
			leafTillerNumberArray1.Add(TillerNumber);
		}
		
		leafTillerNumberArray = new List<double>();
		for (int i = 0; i < leafTillerNumberArray1.Count;i++ ) leafTillerNumberArray.Add(leafTillerNumberArray1[i]);
		

	}

	public static void Init(double SowingDensity)
	{

		double CanopyShootNumber = SowingDensity;
		double AverageShootNumberPerPlant;
		AverageShootNumberPerPlant =1;
		List<double>tilleringProfile = new List<double>();
		tilleringProfile.Add(SowingDensity);
		int TillerNumber;
		TillerNumber = 1;
		
	}
	static void CalcShootNumber(double LeafNumber, double SowingDensity,double TargetFertileShoot)
	{
		int EmergedLeaves = (int)Math.Max(1, Math.Ceiling(LeafNumber - 1));
		int Shoots = Fibonacci(EmergedLeaves);
		double CanopyShootNumber = Math.Min(Shoots * SowingDensity, TargetFertileShoot);
		double AverageShootNumberPerPlant = CanopyShootNumber / SowingDensity;
	}
	static int Fibonacci(int N)
	{
		int a = 0;
		int b = 1;
		for (int i = 0; i < N; i++)
		{
			int temp = a;
			a = b;
			b = temp + b;
		}
		return a;
	}
	
}



